feat: Add all-time contributions support with deduplicated repository count

banu4prasad opened the pull request:
## Summary
Adds support for displaying all-time contributions across all years on GitHub, with automatic deduplication of repositories to show accurate unique contribution counts.

## Changes
- Added `all_time_contribs` parameter to toggle between last year and all-time contributions
- Deduplicated repository counting (counts each repo only once, even with multiple contribution types)
- Increased cache duration for all-time stats.
- Added `ALL_TIME_CONTRIBS` environment variable to enable/disable feature globally

## Usage
Standard (last year only)
?username=YOUR_USERNAME

All-time contributions (deduplicated)
?username=YOUR_USERNAME&all_time_contribs=true

## Environment Variables
- `ALL_TIME_CONTRIBS=true` - Enable all-time contributions feature
- `PAT_1=<github_token>` - GitHub Personal Access Token with `read:user` scope

fixes #2282 

<img width="569" height="109" alt="Screenshot 2025-11-08 at 4 46 26 PM" src="https://github.com/user-attachments/assets/c839186e-23bc-4786-bc03-ce7979ee269f" />
<img width="1040" height="237" alt="Screenshot 2025-11-08 at 4 40 17 PM" src="https://github.com/user-attachments/assets/97073025-b67e-4fb8-8c0b-da5c8f98e9a5" />
<img width="1040" height="237" alt="Screenshot 2025-11-08 at 4 40 05 PM" src="https://github.com/user-attachments/assets/40f564ae-765a-4cbe-bb53-818994f07758" />

banu4prasad committed:
gitignore

banu4prasad committed:
[feat] all_time_contribs

banu4prasad committed:
quick fix

banu4prasad committed:
[fix] allContrib

banu4prasad committed:
[fix] timeout error

banu4prasad committed:
[fix] logger.log

banu4prasad committed:
minor changes

banu4prasad committed:
changes 2

banu4prasad committed:
fixx

banu4prasad committed:
force=F

banu4prasad committed:
trying

banu4prasad committed:
changes

banu4prasad committed:
remove deduplicate

banu4prasad committed:
Merge branch 'anuraghazra:master' into master

banu4prasad committed:
Merge branch 'anuraghazra:master' into master

banu4prasad committed:
changes as per the review

banu4prasad committed:
Merge branch 'anuraghazra:master' into master

banu4prasad committed:
removed unnecessary import

banu4prasad committed:
minor fix

banu4prasad committed:
cache duration

banu4prasad committed:
add envs

banu4prasad committed:
test file

banu4prasad committed:
changes

dargmuesli commented on src/translations.js:
```diff
@@ -351,6 +351,54 @@ const statCardLocales = ({ name, apostrophe }) => {
       no: "Bidro til (i fjor)",
     },
+    "statcard.contribs-alltime": {
+      en: "Contributed to (all time)",
```
What about showing `Contributed to` without `(all time)` to stay in sync with the `Total Commits` design? I'd love it to stay simple.

banu4prasad replied:
i have changed it to `contributed to` in `en` lang

banu4prasad replied:
<img width="467" height="193" alt="Screenshot 2026-02-01 at 1 02 01 AM" src="https://github.com/user-attachments/assets/f751e30f-98ad-4912-ae32-0b50d418675b" />

dargmuesli replied:
Thx! I can't mark this thread as resolved as I don't have the permission in this repo, feel free to do so

banu4prasad committed:
remove unused imports

banu4prasad committed:
reviewed changes

banu4prasad committed:
reviewed changes

banu4prasad committed:
readme changes

banu4prasad committed:
minor changes

banu4prasad committed:
Add support for request options and abort signal in fetchers and retryer

banu4prasad committed:
update version and functionality

banu4prasad committed:
.

google-labs-jules[bot] committed:
🔒 Fix reflected XSS in API status endpoints

Sanitize error messages in `api/status/up.js` and `api/status/pat-info.js` to prevent Reflected XSS.

The original implementation concatenated error messages directly into `res.send()` without proper HTML encoding. This has been addressed by leveraging the existing `encodeHTML` utility from `src/common/html.js` to escape the error messages before sending them.

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

banu4prasad committed:
Merge pull request #1 from banu4prasad/fix-reflected-xss-api-status-12617991897529006981

🔒 Fix Reflected XSS in API Status Endpoints

banu4prasad committed:
Merge branch 'anuraghazra:master' into master

google-labs-jules[bot] committed:
perf: optimize O(N^2) array concatenation in top-languages fetcher

Replaced `curr.languages.edges.concat(acc)` with `acc.push(...curr.languages.edges)` to prevent recreating the accumulator array on each iteration. This reduces time complexity from O(N^2) to O(N).

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

banu4prasad committed:
url.parse() fix

banu4prasad committed:
Merge pull request #2 from banu4prasad/perf/top-languages-reduce-concat-9829857916427544140

⚡ perf: optimize O(N^2) array concatenation in top-languages fetcher

google-labs-jules[bot] committed:
Fix unescaped secondaryMessage XSS vulnerability

This commit escapes the `secondaryMessage` inside `renderError` using `encodeHTML()` to prevent XSS vulnerabilities when generating SVG output.

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
🔒 Fix Potential Reflected XSS in PAT info and Status endpoints

Updates `api/status/pat-info.js` and `api/status/up.js` to return proper structured JSON using `res.json()` rather than sending unescaped strings or stringified JSON via `res.send()`. This explicitly sets the correct Content-Type ensuring browsers won't execute malicious error messages as HTML. Updates tests accordingly.

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
⚡ Optimize object property access in top languages fetcher

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
⚡ perf: Optimize hidden stats array lookups with Set

Changed the \`hide\` array filtering within \`renderStatsCard\` to use a \`Set\`.
This changes the O(n) \`.includes()\` lookup to an O(1) \`.has()\` lookup.
Applied to both stat items filtering and accessibility labels.

Performance baseline measured using \`tests/bench/stats.bench.js\`:
Before: avg ~32.5µs
After: avg ~26.9µs

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

banu4prasad committed:
Merge pull request #5 from banu4prasad/optimize-top-langs-property-access-11153830499985198609

⚡ Optimize object property access in top languages fetcher

banu4prasad committed:
Merge pull request #6 from banu4prasad/perf-optimization-hide-includes-410756806148804890

⚡ Optimize hidden stats array lookups with Set

google-labs-jules[bot] committed:
🔒 Format error responses to JSON objects in PAT info and Uptime endpoints

Updates `api/status/pat-info.js` and `api/status/up.js` error handling to explicitly return a JSON object using `res.json({ error: ... })` instead of sending strings through `res.send()`. This enforces strict API consistency and mitigates any theoretical HTML rendering issues, resolving potential XSS concerns. Note this is a breaking change for existing clients directly parsing the raw string error response. Tests have been updated to cover the new outer catch payload structure.

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
🔒 Format error responses to JSON objects in PAT info and Uptime endpoints

Updates `api/status/pat-info.js` and `api/status/up.js` error handling to explicitly return a JSON object using `res.json({ error: ... })` instead of sending strings through `res.send()`. This enforces strict API consistency and corrects the API behavior since the endpoint explicitly sets `Content-Type: application/json`. Tests have been updated to cover the new outer catch payload structure.

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

banu4prasad committed:
Merge pull request #4 from banu4prasad/fix/security-xss-pat-info-6799410205025217312

🔒 Fix Reflected XSS Vulnerability in PAT Info and Uptime Endpoints

google-labs-jules[bot] committed:
Fix unescaped secondaryMessage XSS vulnerability

This commit escapes the `secondaryMessage` inside `renderError` using `encodeHTML()` to prevent XSS vulnerabilities when generating SVG output. Also avoids double-escaping regression.

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
Fix unescaped secondaryMessage XSS vulnerability

This commit escapes the `secondaryMessage` inside `renderError` using `encodeHTML()` to prevent XSS vulnerabilities when generating SVG output. Also avoids double-escaping regression and adds robust tests.

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

banu4prasad committed:
Merge pull request #3 from banu4prasad/security-fix-unescaped-secondary-message-14295917276723695873

🔒 Fix unescaped secondaryMessage XSS vulnerability in renderError

google-labs-jules[bot] committed:
perf: optimize batch processing with worker pool in `all-time-contributions.js`

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
perf: optimize batch processing with worker pool in `all-time-contributions.js`

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
perf: optimize batch processing with worker pool in `all-time-contributions.js`

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
perf: optimize batch processing with worker pool in `all-time-contributions.js`

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

google-labs-jules[bot] committed:
perf: optimize batch processing with worker pool in `all-time-contributions.js`

Co-authored-by: banu4prasad <113224132+banu4prasad@users.noreply.github.com>

banu4prasad committed:
Merge pull request #7 from banu4prasad/jules-optimize-process-batched-219987405495775010

⚡ perf: optimize batch processing in all-time contributions with worker pool

dargmuesli commented:
@banu4prasad why was this closed? :eyes:

banu4prasad commented:
> @banu4prasad why was this closed? 👀

@dargmuesli cuz i have edited too many files which was not related to the PR

dargmuesli commented:
@banu4prasad do you plan on creating a more concise PR or what is your thoughts on moving this forwards at this point? Thank you for your work on this so far 💚
